### PR TITLE
[@mantine/core] Stepper: Add preventClickAhead to prevent clicking beyond last reached step

### LINKED ISSUE #555: [Feature Request] Stepper should have prop to prevent clicking beyond last reached step.
### Usage example, including component/hook API

For certain workflows you'd want your customers to be able to navigate freely throughout the steps they've already accessed, but not further ahead. It would be nice to have that configurable via a prop:

```typescript
const [page, setPage] = useState<number>(0);

// ...

<Stepper active={page} onStepClick={setPage} preventClickAhead={true}>
    <Stepper.Step>
        Welcome Instructions
    </Stepper.Step>
    <Stepper.Step>
        First Form Page
    </Stepper.Step>
    <Stepper.Step>
        Second Form Page
    </Stepper.Step>
    <Stepper.Step>
        Confirmation
    </Stepper.Step>
</Stepper>

```

In the example above, if the customer reached the First Form Page, they would be able to click to jump between the Welcome Instructions step and the First Form Page step, but not yet to the Second Form Page. Only once they complete certain actions (for example filling out the form) and click on the Next Page button at the bottom (separately rendered / implemented) would they be able to move on.

### Possible implementation – describe how the feature can be implemented

A possible implementation of this:
1. The stepper would store either the latest step reached, or a list of all steps that have been accessed
2. Modify the `allowStepClick` to something like `allowStepClick={typeof onStepClick === 'function' && (preventClickAhead ? accessedSteps.indexOf(index) != -1)} : true)`

### Do you want to contribute this feature and create a pull request

_No response_

### PR DESCRIPTION
**Pull request by LucasLemanowicz**
Fixes #555

### CODE REVIEW

File: src/mantine-core/src/components/Stepper/Step/Step.tsx [hunk 1]
```diff
@@ -95,6 +99,12 @@ export const Step = forwardRef<HTMLButtonElement, StepProps>(
     const _iconSize = theme.fn.size({ size, sizes: defaultIconSizes });
     const _icon = state === 'stepCompleted' ? null : state === 'stepProgress' ? progressIcon : icon;
 
+    useEffect(() => {
+      if (['stepProgress', 'stepCompleted'].includes(state)) {
+        onStepReached && onStepReached();
```
Comment (rtivital) [inline src/mantine-core/src/components/Stepper/Step/Step.tsx:104]: I think there is no need for this prop as Stepper is always controlled

Comment (LucasLemanowicz) [reply src/mantine-core/src/components/Stepper/Step/Step.tsx]: Good point, fixed.

**Comment by rtivital** [comment]
@LucasLemanowicz please do the following to merge this PR:
- resolve conflicts with next-minor branch
- add new demo that will showcase this feature
- add documentation with a demo that will tell how to use this feature

**Comment by LucasLemanowicz** [comment]
Thanks, I will take care of this in the next week or two!

**Comment by rtivital** [comment]
I've come up with a more flexible solution for this with `allowStepSelect` prop on Step component. So closing this PR since it's no longer relevant.

```tsx
import { Stepper, Button, Group } from '@mantine/core';

function Demo() {
  const [active, setActive] = useState(1);
  const nextStep = () => setActive((current) => (current < 3 ? current + 1 : current));
  const prevStep = () => setActive((current) => (current > 0 ? current - 1 : current));

  return (
    <>
      <Stepper active={active} onStepClick={setActive} breakpoint="sm">
        <Stepper.Step label="Fist step" description="Create an account" allowStepSelect={active > 0}>
          Step 1 content: Create an account
        </Stepper.Step>
        <Stepper.Step label="Second step" description="Verify email" allowStepSelect={active > 1}>
          Step 2 content: Verify email
        </Stepper.Step>
        <Stepper.Step label="Final step" description="Get full access" allowStepSelect={active > 2}>
          Step 3 content: Get full access
        </Stepper.Step>
        <Stepper.Completed>
          Completed, click back button to get to previous step
        </Stepper.Completed>
      </Stepper>

      <Group position="center" mt="xl">
        <Button variant="default" onClick={prevStep}>Back</Button>
        <Button onClick={nextStep}>Next step</Button>
      </Group>
    </>
  );
}
```

**Pull request closed without merging**